File: Solution_2463/Solution.java
class Solution {
    private long[][] f;
    private List<Integer> robot;
    private int[][] factory;

    public long minimumTotalDistance(List<Integer> robot, int[][] factory) {
        // Sort robots and factories by position to enable nearest pairing
        Collections.sort(robot);
        Arrays.sort(factory, (a, b) -> a[0] - b[0]);

        this.robot = robot;
        this.factory = factory;
        f = new long[robot.size()][factory.length];
        for (int i = 0; i < robot.size(); i++) {
            Arrays.fill(f[i], -1);  // Initialize with -1 to denote uncomputed values
        }

        return dfs(0, 0);
    }

    private long dfs(int i, int j) {
        // All robots are repaired
        if (i == robot.size()) {
            return 0;
        }
        // No more factories to repair remaining robots
        if (j == factory.length) {
            return Long.MAX_VALUE / 1000;
        }
        // Return cached result if computed
        if (f[i][j] != -1) {
            return f[i][j];
        }

        // Case 1: Skip current factory
        long ans = dfs(i, j + 1);
        
        // Case 2: Use current factory to repair up to `limit` robots
        long t = 0;
        for (int k = 0; k < factory[j][1]; ++k) {
            if (i + k == robot.size()) {
                break;
            }
            // Accumulate distance for assigning robot[i + k] to factory[j]
            t += Math.abs(robot.get(i + k) - factory[j][0]);
            ans = Math.min(ans, t + dfs(i + k + 1, j + 1));
        }

        f[i][j] = ans;  // Cache result
        return ans;
    }
}

File: Solution_2463/Solution.js
var minimumTotalDistance = function(robot, factory) {
    // Sort robots and factories by their positions to optimize pairing
    robot.sort((a, b) => a - b);
    factory.sort((a, b) => a[0] - b[0]);

    // Initialize memoization table with undefined to indicate uncomputed states
    const memo = Array.from({ length: robot.length }, () => Array(factory.length).fill(undefined));

    // Recursive DFS function to calculate minimum distance
    function dfs(i, j) {
        // Base Case: All robots have been repaired
        if (i === robot.length) return 0;
        
        // Base Case: No more factories left, but robots still need repairs
        if (j === factory.length) return Number.MAX_SAFE_INTEGER / 1000;

        // Check if result is already computed
        if (memo[i][j] !== undefined) return memo[i][j];

        // Option 1: Skip current factory and try the next one
        let ans = dfs(i, j + 1);

        // Option 2: Use the current factory to repair up to its limit of robots
        let t = 0; // Tracks accumulated distance for pairing robots with this factory
        for (let k = 0; k < factory[j][1]; ++k) { // Loop up to the factory's repair limit
            if (i + k === robot.length) break; // Break if there are no more robots left

            // Calculate the distance for assigning robot[i + k] to factory[j]
            t += Math.abs(robot[i + k] - factory[j][0]);

            // Update ans with the minimum distance found between skipping or using this factory
            ans = Math.min(ans, t + dfs(i + k + 1, j + 1));
        }

        // Store the result in the memoization table for this robot-factory pair
        memo[i][j] = ans;

        // Return the minimum distance for the current state
        return ans;
    }

    // Start the DFS search with the first robot and the first factory
    return dfs(0, 0);
};
